pcsx2-qt/Debugger/DisassemblyWidget.cpp

```diff
@@ -0,0 +1,699 @@
+	m_contextMenu->addAction(action = new QAction("Assemble new Instruction(s)", this));
+	connect(action, SIGNAL(triggered()), SLOT(contextAssembleInstruction()));
+	m_contextMenu->addSeparator();
+	m_contextMenu->addAction(action = new QAction("Run to Cursor", this));
+	connect(action, SIGNAL(triggered()), SLOT(contextRunToCursor()));
+	m_contextMenu->addAction(action = new QAction("Jump to Cursor", this));
+	connect(action, SIGNAL(triggered()), SLOT(contextJumpToCursor()));
+	m_contextMenu->addAction(action = new QAction("Toggle Breakpoint", this));
+	connect(action, SIGNAL(triggered()), SLOT(contextToggleBreakpoint()));
+	m_contextMenu->addAction(action = new QAction("Follow Branch", this));
+	connect(action, SIGNAL(triggered()), SLOT(contextFollowBranch()));
+	m_contextMenu->addSeparator();
+	m_contextMenu->addAction(action = new QAction("Go to Address", this)); // Needs a better name imo.
+	connect(action, SIGNAL(triggered()), SLOT(contextGoToAddress()));
+	m_contextMenu->addAction(action = new QAction("Go to in Memory View", this));
+	connect(action, &QAction::triggered, this, [this]() { emit gotoInMemory(m_selectedAddressStart); });
+	m_contextMenu->addSeparator();
+	m_contextMenu->addAction(action = new QAction("Add Function", this));
+	connect(action, SIGNAL(triggered()), SLOT(contextAddFunction()));
+	m_contextMenu->addAction(action = new QAction("Rename Function", this));
+	connect(action, SIGNAL(triggered()), SLOT(contextRenameFunction()));
+	m_contextMenu->addAction(action = new QAction("Remove Function", this));
+	connect(action, SIGNAL(triggered()), SLOT(contextRemoveFunction()));
+}
+
+void DisassemblyWidget::contextCopyAddress()
+{
+	QGuiApplication::clipboard()->setText(FetchSelectionInfo(SelectionInfo::ADDRESS));
+}
+
+void DisassemblyWidget::contextCopyInstructionHex()
+{
+	QGuiApplication::clipboard()->setText(FetchSelectionInfo(SelectionInfo::INSTRUCTIONHEX));
+}
+
+void DisassemblyWidget::contextCopyInstructionText()
+{
+	QGuiApplication::clipboard()->setText(FetchSelectionInfo(SelectionInfo::INSTRUCTIONTEXT));
+}
+
+void DisassemblyWidget::contextAssembleInstruction()
+{
+	if (!m_cpu->isCpuPaused())
+	{
+		QMessageBox::warning(this, "Assemble Error", "Unable to change assembly while core is running");
+		return;
+	}
+
+	DisassemblyLineInfo line;
+	bool ok;
+	m_disassemblyManager.getLine(m_selectedAddressStart, false, line);
+	QString instruction = QInputDialog::getText(this, "Assemble Instruction", "",
+		QLineEdit::Normal, QString("%1 %2").arg(line.name.c_str()).arg(line.params.c_str()), &ok);
+
+	if (!ok)
+		return;
+
+	u32 encodedInstruction;
+	std::string errorText;
+	bool valid = MipsAssembleOpcode(instruction.toLocal8Bit().constData(), m_cpu, m_selectedAddressStart, encodedInstruction, errorText);
+
+	if (!valid)
+	{
+		QMessageBox::warning(this, "Assemble Error", QString::fromStdString(errorText));
+		return;
+	}
+	else
+	{
+		for (u32 i = m_selectedAddressStart; i <= m_selectedAddressEnd; i += 4)
```

Comment:
Probably want to execute this on the CPU thread, that way any invalidations don't occur on the UI thread. (I'm assuming `write32()` doesn't do this internally, if it does, disregard)